### PR TITLE
fix: Recipe update skip ignored files

### PR DESCRIPTION
**Pull request by maxime-aknin**
I ran into an issue with the `composer recipes:update` command. While trying to update `doctrine/doctrine-bundle` recipe from 2.0 to 2.4, the command ouput stated that the recipe was updated with conflicts, but there were no actual changes made in files (except the symfony lock file):
Screen Shot 2022-03-15 at 5 15 21 PM

What's happening is that this recipe update contains changes that are applied to `phpunit.xml` by ` Symfony\Flex\Configurator\EnvConfigurator`, but this file is also added to `gitignore` by `phpunit/phpunit` [recipe](https://github.com/symfony/recipes/blob/2d1ba2cf32556d9c01ff7bd05bb02ec9bfb44e5c/phpunit/phpunit/9.3/manifest.json#L8)

The fact that there is an ignored file in the recipe patch makes the `git apply` command in [RecipePatcher](https://github.com/symfony/flex/blob/b66b0dd8f5099002f4863aa0bba0c3e0c9e8523b/src/Update/RecipePatcher.php#L55)  fail without applying the patch, and this failure is not detected by the rest of the function.
Screen Shot 2022-03-15 at 5 31 10 PM

My proposal is to filter out ignored files from the beginning of the `generatePatch` function to fix this issue.

### CODE REVIEW

**Review by dunglas — APPROVED** [review]

File: src/Update/RecipePatcher.php [hunk 1]
```diff
@@ -76,13 +76,16 @@ public function applyPatch(RecipePatch $patch): bool
...
+	    $ignoredFiles = $this->getIgnoredFiles(array_keys($originalFiles) + array_keys($newFiles));
+
+	    // null implies "file does not exist"
+	    $originalFiles = array_filter($originalFiles, function ($file, $fileName) use ($ignoredFiles) {
+		    return null !== $file && !in_array($fileName, $ignoredFiles);
+	    }, ARRAY_FILTER_USE_BOTH);
+
+	    $newFiles = array_filter($newFiles, function ($file, $fileName) use ($ignoredFiles) {
+		    return null !== $file && !in_array($fileName, $ignoredFiles);
+	    }, ARRAY_FILTER_USE_BOTH);
```
Comment (weaverryan) [inline src/Update/RecipePatcher.php]: Looks like some indentation problems - maybe some tab characters slipped in?

File: src/Update/RecipePatcher.php [hunk 2]
```diff
@@ -76,13 +76,16 @@ public function applyPatch(RecipePatch $patch): bool
...
-            return null !== $file;
-        });
-        $newFiles = array_filter($newFiles, function ($file) {
-            return null !== $file;
-        });
+	    $ignoredFiles = $this->getIgnoredFiles(array_keys($originalFiles) + array_keys($newFiles));
+
+	    // null implies "file does not exist"
+	    $originalFiles = array_filter($originalFiles, function ($file, $fileName) use ($ignoredFiles) {
+		    return null !== $file && !in_array($fileName, $ignoredFiles);
```
Comment (weaverryan) [inline src/Update/RecipePatcher.php]:
```suggestion
		    return null !== $file && !\in_array($fileName, $ignoredFiles);
```

File: src/Update/RecipePatcher.php [hunk 3]
```diff
@@ -76,13 +76,16 @@ public function applyPatch(RecipePatch $patch): bool
...
+	    }, ARRAY_FILTER_USE_BOTH);
+
+	    $newFiles = array_filter($newFiles, function ($file, $fileName) use ($ignoredFiles) {
+		    return null !== $file && !in_array($fileName, $ignoredFiles);
```
Comment (weaverryan) [inline src/Update/RecipePatcher.php]:
```suggestion
		    return null !== $file && !\in_array($fileName, $ignoredFiles);
```

File: src/Update/RecipePatcher.php [hunk 4]
```diff
@@ -223,4 +226,12 @@ private function getBlobPath(string $hash): string
...
         return '.git/objects/'.$hashStart.'/'.$hashEnd;
     }
+
+	private function getIgnoredFiles(array $fileNames): array
+	{
+		$args = implode(' ', array_map([ProcessExecutor::class, 'escape'], $fileNames));
+		$output = '';
+		$this->processExecutor->execute(sprintf('git check-ignore %s', $args), $output, $this->rootDir);
+		return $this->processExecutor->splitLines($output);
+	}
```
Comment (weaverryan) [inline src/Update/RecipePatcher.php]: Also here: tabs should be spaces :)

File: src/Update/RecipePatcher.php [hunk 5]
```diff
@@ -223,4 +226,12 @@ private function getBlobPath(string $hash): string
 
         return '.git/objects/'.$hashStart.'/'.$hashEnd;
     }
+
+	private function getIgnoredFiles(array $fileNames): array
+	{
+		$args = implode(' ', array_map([ProcessExecutor::class, 'escape'], $fileNames));
+		$output = '';
+		$this->processExecutor->execute(sprintf('git check-ignore %s', $args), $output, $this->rootDir);
+		return $this->processExecutor->splitLines($output);
```
Comment (weaverryan) [inline src/Update/RecipePatcher.php]: New line between `$this->processExecutor...` and the `return` statement.

File: tests/Update/RecipePatcherTest.php [hunk 6]
```diff
@@ -38,13 +38,31 @@ public function testGeneratePatch(array $originalFiles, array $newFiles, string
...
+                file_put_contents(FLEX_TEST_DIR.'/'.$file, $contents);
+            }
+        }
+
+        // make sure the test directory is a git repo
+        (new Process(['git', 'init'], FLEX_TEST_DIR))->mustRun();
+        (new Process(['git', 'config', 'commit.gpgsign', 'false'], FLEX_TEST_DIR))->mustRun();
+        (new Process(['git', 'config', 'user.name', '"Flex Updater"'], FLEX_TEST_DIR))->mustRun();
+        (new Process(['git', 'config', 'user.email', '""'], FLEX_TEST_DIR))->mustRun();
+        if (!empty($originalFiles)) {
```
Comment (weaverryan) [inline tests/Update/RecipePatcherTest.php]:
```suggestion
        if (0 !== \count($originalFiles)) {
```

File: tests/Update/RecipePatcherTest.php [hunk 7]
```diff
@@ -38,13 +38,31 @@ public function testGeneratePatch(array $originalFiles, array $newFiles, string
...
         foreach ($originalFiles as $file => $contents) {
             touch(FLEX_TEST_DIR.'/'.$file);
+            if ($file === '.gitignore') {
+                file_put_contents(FLEX_TEST_DIR.'/'.$file, $contents);
+            }
+        }
+
+        // make sure the test directory is a git repo
+        (new Process(['git', 'init'], FLEX_TEST_DIR))->mustRun();
+        (new Process(['git', 'config', 'commit.gpgsign', 'false'], FLEX_TEST_DIR))->mustRun();
```
Comment (weaverryan) [inline tests/Update/RecipePatcherTest.php]: Is this strictly needed?

File: tests/Update/RecipePatcherTest.php [hunk 8]
```diff
@@ -38,13 +38,31 @@ public function testGeneratePatch(array $originalFiles, array $newFiles, string
...
+            (new Process(['git', 'commit', '-m', '"original files"'], FLEX_TEST_DIR))->mustRun();
         }
 
         $patcher = new RecipePatcher(FLEX_TEST_DIR, $this->createMock(IOInterface::class));
 
         $patch = $patcher->generatePatch($originalFiles, $newFiles);
         $this->assertSame($expectedPatch, rtrim($patch->getPatch(), "\n"));
 
+        // when testing ignored files the patch is empty
+        if (empty($expectedPatch)) {
```
Comment (weaverryan) [inline tests/Update/RecipePatcherTest.php]:
```suggestion
        if ('' === $expectedPatch) {
```

**Review by weaverryan — COMMENTED** [review]
Thank you for this @maxime-aknin! I thought I had all the edge cases covered. I was wrong :).

I've left some comments - mostly minutiae coding standard changes.

Cheers!

**Comment by weaverryan** [comment]
Oh, and for the coding standards, I should have just pointed you at fabbot (one of the CI checks): https://fabbot.io/report/symfony/flex/885/d84cd84094be56d55cfd5db080db4c40a6dcce16 - you can fix everything automatically from there.

File: tests/Update/RecipePatcherTest.php (hunk 7 above)
Comment (maxime-aknin) [reply tests/Update/RecipePatcherTest.php]: Yeah I'm not sure about this, creating the git repo failed in the ci environnement on my first try so I copied this from [RecipePatcher](https://github.com/symfony/flex/blob/d84cd84094be56d55cfd5db080db4c40a6dcce16/src/Update/RecipePatcher.php#L113) to mimic the way the git repo is created. But this command about gpgsign might not be needed. I'll try and remove it and see if the ci tests still pass :)

Comment (maxime-aknin) [reply tests/Update/RecipePatcherTest.php]: Seems like it passed !

**Comment by maxime-aknin** [comment]
Hello @weaverryan 👋

That's ok I think every code style problem is now fixed. Sorry I didn't take the time to properly check the code standards before submitting. Thanks for the feedback and for the great symfonycast 🤘

Cheers !

**Comment by fabpot** [comment]
@weaverryan Can you check now that the code has been updated now?

**Review by maxhelias — APPROVED** [review]

**Review by weaverryan — APPROVED** [review]
Looks great 👍

**Comment by fabpot** [comment]
Thank you @maxime-aknin.
